File: backend/services/cache_service.py
"""
Cache service layer for consistent cache access and invalidation.

This module provides a service layer that wraps the cache manager and ensures
that write operations properly invalidate related cache entries.
"""

import logging
from typing import Any, Callable, Dict, Optional, TypeVar

from infrastructure.cache import (
    CacheManager,
    agent_config_key,
    agent_object_key,
    chatting_agents_key,
    get_cache,
    room_agents_key,
    room_messages_key,
    room_object_key,
)

logger = logging.getLogger("CacheService")

T = TypeVar("T")


class CacheService:
    """
    Service layer for cache operations with automatic invalidation.

    This service provides a high-level interface to the cache manager
    and ensures that write operations properly invalidate related cache entries.
    """

    def __init__(self, cache_manager: Optional[CacheManager] = None):
        """
        Initialize cache service.

        Args:
            cache_manager: Optional CacheManager instance (uses global if not provided)
        """
        self._cache = cache_manager or get_cache()

    # Read operations

    def get(self, key: str) -> Optional[Any]:
        """Get value from cache."""
        return self._cache.get(key)

    def set(self, key: str, value: Any, ttl_seconds: float = 60):
        """Set value in cache with TTL."""
        self._cache.set(key, value, ttl_seconds)

    def get_or_set(self, key: str, factory: Callable[[], T], ttl_seconds: float = 60) -> T:
        """Get value from cache or compute if not found."""
        return self._cache.get_or_set(key, factory, ttl_seconds)

    async def get_or_set_async(self, key: str, factory: Callable[[], Any], ttl_seconds: float = 60) -> Any:
        """Async version of get_or_set."""
        return await self._cache.get_or_set_async(key, factory, ttl_seconds)

    # Invalidation operations

    def invalidate(self, key: str) -> bool:
        """Invalidate a specific cache key."""
        return self._cache.invalidate(key)

    def invalidate_pattern(self, pattern: str):
        """Invalidate all keys matching a pattern (prefix match)."""
        self._cache.invalidate_pattern(pattern)

    def invalidate_agent(self, agent_id: int):
        """
        Invalidate all cache entries related to an agent.

        This should be called when:
        - Agent is created, updated, or deleted
        - Agent configuration is reloaded
        """
        keys = [agent_object_key(agent_id), agent_config_key(agent_id)]
        for key in keys:
            self._cache.invalidate(key)
        logger.debug(f"Invalidated cache for agent {agent_id}")

    def invalidate_room(self, room_id: int):
        """
        Invalidate all cache entries related to a room.

        This should be called when:
        - Room is created, updated, or deleted
        - Room settings change (paused, max_interactions)
        """
        keys = [room_object_key(room_id), room_agents_key(room_id), chatting_agents_key(room_id)]
        for key in keys:
            self._cache.invalidate(key)
        # Also invalidate message cache pattern
        self._cache.invalidate_pattern(room_messages_key(room_id))
        logger.debug(f"Invalidated cache for room {room_id}")

    def invalidate_room_agents(self, room_id: int):
        """
        Invalidate room agents cache.

        This should be called when:
        - Agents are added to or removed from a room
        """
        self._cache.invalidate(room_agents_key(room_id))
        logger.debug(f"Invalidated room agents cache for room {room_id}")

    def invalidate_room_messages(self, room_id: int):
        """
        Invalidate message cache for a room.

        This should be called when:
        - New messages are created
        """
        self._cache.invalidate_pattern(room_messages_key(room_id))
        logger.debug(f"Invalidated message cache for room {room_id}")

    # Maintenance operations

    def cleanup_expired(self):
        """Remove all expired entries from cache."""
        self._cache.cleanup_expired()

    def clear(self):
        """Clear all cache entries."""
        self._cache.clear()

    # Statistics and monitoring

    def get_stats(self) -> Dict[str, int]:
        """
        Get cache statistics.

        Returns:
            Dictionary with hits, misses, hit_rate, size, invalidations
        """
        return self._cache.get_stats()

    def log_stats(self):
        """Log current cache statistics."""
        self._cache.log_stats()


# Global cache service instance
_cache_service: Optional[CacheService] = None


def get_cache_service() -> CacheService:
    """Get the global cache service instance."""
    global _cache_service
    if _cache_service is None:
        _cache_service = CacheService()
    return _cache_service


def reset_cache_service():
    """Reset the cache service instance (useful for testing)."""
    global _cache_service
    _cache_service = None

File: backend/src/services/cache-service.ts
/**
 * Facade over the in-memory request cache, adding one thing to `CacheManager`:
 * named invalidation groups, so a caller that changed a room need not remember
 * which four keys a room contributes. `infrastructure/cache.ts` owns TTL, LRU
 * and single-flight.
 *
 * **This is not the only invalidator.** `crud/cached.ts` has its own, and the
 * CRUD write paths call *those*; the key sets match except as noted on
 * {@link CacheService.invalidateRoom}.
 */

import {
  agentConfigKey,
  agentObjectKey,
  chattingAgentsKey,
  getCache,
  roomAgentsKey,
  roomMessagesKey,
  roomObjectKey,
  type CacheManager,
  type CacheStats,
} from '@/infrastructure/cache'
import { getLogger } from '@/infrastructure/logging/logger'

const logger = getLogger('CacheService')

export class CacheService {
  private readonly cache: CacheManager

  /** Defaults to the process-wide cache; an explicit one keeps tests isolated. */
  constructor(cacheManager: CacheManager = getCache()) {
    this.cache = cacheManager
  }

  get<T = unknown>(key: string): T | undefined {
    return this.cache.get<T>(key)
  }

  set(key: string, value: unknown, ttlSeconds = 60): void {
    this.cache.set(key, value, ttlSeconds)
  }

  getOrSet<T = unknown>(key: string, factory: () => T, ttlSeconds = 60): T {
    return this.cache.getOrSet<T>(key, factory, ttlSeconds)
  }

  getOrSetAsync<T>(key: string, factory: () => Promise<T>, ttlSeconds = 60): Promise<T> {
    return this.cache.getOrSetAsync<T>(key, factory, ttlSeconds)
  }

  invalidate(key: string): boolean {
    return this.cache.invalidate(key)
  }

  /** Prefix match, not a glob, despite the name. */
  invalidatePattern(pattern: string): void {
    this.cache.invalidatePattern(pattern)
  }

  cleanupExpired(): void {
    this.cache.cleanupExpired()
  }

  clear(): void {
    this.cache.clear()
  }

  /**
   * Cache counters. `/debug/cache/stats` is a frozen wire shape with snake_case
   * keys (`total_requests`, `hit_rate`) while {@link CacheStats} is camelCase,
   * so that route must map the fields rather than serialise this directly.
   */
  getStats(): CacheStats {
    return this.cache.getStats()
  }

  logStats(): void {
    this.cache.logStats()
  }

  invalidateAgent(agentId: number): void {
    this.cache.invalidate(agentObjectKey(agentId))
    this.cache.invalidate(agentConfigKey(agentId))
    logger.debug(`Invalidated cache for agent ${agentId}`)
  }

  /**
   * Clears one key more than `crud/cached.ts::invalidateRoomCache`:
   * `chatting_agents:{id}`. The two are called from different places and the
   * extra key costs only a few saved queries.
   *
   * The message sweep is a prefix match on `room_messages:{id}` with no trailing
   * separator, so invalidating room 1 also clears rooms 10 and 100 — the same
   * quirk `cached.ts` has, and it stays for the reason noted there.
   */
  invalidateRoom(roomId: number): void {
    this.cache.invalidate(roomObjectKey(roomId))
    this.cache.invalidate(roomAgentsKey(roomId))
    this.cache.invalidate(chattingAgentsKey(roomId))
    this.cache.invalidatePattern(roomMessagesKey(roomId))
    logger.debug(`Invalidated cache for room ${roomId}`)
  }

  invalidateRoomAgents(roomId: number): void {
    this.cache.invalidate(roomAgentsKey(roomId))
    logger.debug(`Invalidated room agents cache for room ${roomId}`)
  }

  invalidateRoomMessages(roomId: number): void {
    this.cache.invalidatePattern(roomMessagesKey(roomId))
    logger.debug(`Invalidated message cache for room ${roomId}`)
  }
}

let cachedService: CacheService | null = null

/** Lazy so {@link resetCacheService} can hand the next caller a service built
 * against a different manager. */
export function getCacheService(): CacheService {
  if (cachedService === null) cachedService = new CacheService()
  return cachedService
}

/** Drop the singleton. Test-only. */
export function resetCacheService(): void {
  cachedService = null
}
